### PR TITLE
Add a custom test harness and an example crate that shows how to use it

### PR DESCRIPTION
**Pull request by Avi-D-coder**
This PR enables running tests inside of valida.
I have added a tests utilizing this functionality, and a example crate that shows how a user of this crate would use the test harness.

Most Rust tests are supported with minimal changes, however examples, benchmarks, and dynamic tests are not. Their is a clear path to supporting examples benchmarks with the addition of a `cargo valida` subcommand to pass test options like `--examples=` to the runner.

### CODE REVIEW

**Comment by Avi-D-coder** [comment]
> Thanks Avi! Great docs! I got this:
> 
> ```
> ~/code/lita/12Sep/valida-rs/examples/testing$ cargo test
>    Compiling entrypoint v0.1.0 (/home/marty/code/lita/12Sep/valida-rs)
>    Compiling testing v0.1.0 (/home/marty/code/lita/12Sep/valida-rs/examples/testing)
>     Finished `test` profile [unoptimized + debuginfo] target(s) in 0.20s
>      Running unittests src/lib.rs (/home/marty/code/lita/12Sep/valida-rs/target/debug/deps/testing-899dd6b78745721a)
> Building tests for valida
> Running tests host.
> running 3 tests
> test test_add ... ok
> test test_add on valida ... FAILED
> 
> 
> test test_add failure message: Failed to start test process: No such file or directory (os error 2)
> 
> 
> thread 'main' panicked at examples/testing/src/lib.rs:13:5:
> assertion `left == right` failed
>   left: 4
>  right: 5
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> test test_fail ... ok
> test test_fail on valida ... FAILED
> 
> 
> test test_fail failure message: Failed to start test process: No such file or directory (os error 2)
> 
> 
> test test_ignore ... ignored
> 
> test result: ok
> on native:      2 passed; 0 failed
> on valida:      0 passed; 2 failed
> 1 ignored;
> 0 unsupported
> error: test failed, to rerun pass `--lib`
> 
> Caused by:
>   process didn't exit successfully: `/home/marty/code/lita/12Sep/valida-rs/target/debug/deps/testing-899dd6b78745721a` (exit status: 1)
> note: test exited abnormally; to see the full output pass --nocapture to the harness.
> ```

One thing I assume is that valida is installed in $PATH.


Does which valida return anything?

**Comment by thealmarty** [comment]
@Avi-D-coder I got `$ which valida
which: no valida in ...` is there a way to make this not dependent on valida being in $PATH? Or I guess we can just document the requirement? I did also try to add some better error msg here https://github.com/lita-xyz/valida-rs/tree/thealmarty-review feel free to pick and choose or ignore.

**Comment by Avi-D-coder** [comment]
@thealmarty ultimately we have to find valida somewhere so docs is probably the right solution. This will also be fixed once we have a cargo subcommand.

I don't have a work computer with me on vacation, but I will add docs next Monday.

does it work for you if you cargo install valida or otherwise add it to your path. Note I would not expect the valida-shell bash script to worm because it does other stuff too.

**Comment by thealmarty** [comment]
I've rebased and updated some names and ran and got this:

```
~/code/lita/12Sep/valida-rs/examples/testing$ export PATH=$PATH:/home/marty/code/lita/12Sep/valida-toolchain/valida-vm/target/release/
09:50:03 ~/code/lita/12Sep/valida-rs/examples/testing$ cargo test
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running unittests src/lib.rs (/home/marty/code/lita/12Sep/valida-rs/target/debug/deps/testing-43b0810baba8b753)
Building tests for valida
Running tests host.
running 3 tests
test test_add ... ok
test test_add on valida ... ok
thread 'main' panicked at examples/testing/src/lib.rs:13:5:
assertion `left == right` failed
  left: 4
 right: 5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_fail ... ok
test test_fail on valida ... ok
test test_ignore ... ignored

test result: ok
on native:      2 passed; 0 failed
on valida:      2 passed; 0 failed
1 ignored;
0 unsupported
     Running unittests src/main.rs (/home/marty/code/lita/12Sep/valida-rs/target/debug/deps/testing-fa7dc5c95301aa7e)
Building tests for valida
Running tests host.
running 3 tests
thread 'main' panicked at examples/testing/src/main.rs:25:5:
assertion `left == right` failed
  left: 4
 right: 5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test test_fail ... ok
test test_fail on valida ... ok
test test_ignore ... ignored
test test_pass ... ok
test test_pass on valida ... ok

test result: ok
on native:      2 passed; 0 failed
on valida:      2 passed; 0 failed
1 ignored;
0 unsupported
     Running tests/test.rs (/home/marty/code/lita/12Sep/valida-rs/target/debug/deps/test-80c3042ad05597aa)
Building tests for valida
Running tests host.
running 1 tests
test test_integration ... ok
test test_integration on valida ... ok

test result: ok
on native:      1 passed; 0 failed
on valida:      1 passed; 0 failed
0 ignored;
0 unsupported
   Doc-tests testing

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

There's a few panics but looking better!

**Comment by Avi-D-coder** [comment]
@thealmarty those panics are the should_panic tests on the host.

Since it's confusing I can log a message before and after each host should_panic test saying to ignore them.

long term I have a todo about running the host tests in a different process to capture the output like I do for valida tests.

**Comment by thealmarty** [comment]
OK I see. Yeah or at least printing out that it's the output of should_panic test.

So, overall, improved documentation, error reporting (explicit error messages instead of unwrap) , and output layout, otherwise AOK. Thanks!

**Comment by Avi-D-coder** [comment]
@thealmarty could I get another review.

**Comment by Avi-D-coder** [comment]
I'd love to break the test_utils file up into a few modules, but I'll save that for a future work.

File: src/test_utils.rs [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+
+    loop {
+        receive_child_stdout(&mut stdout_buffer);
+
+        let search_end = stdout_buffer
+            .len()
+            .checked_sub(MAGIC_TERMINATOR.len())
+            .unwrap_or(searched_cursor);
+
+        let paniced_with_magic_terminator = (searched_cursor..search_end)
```
Comment (thealmarty) [inline src/test_utils.rs:421]: typo `panicked` and on L425 as well.

File: examples/testing/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[package]
```
Comment (thealmarty) [inline examples/testing/Cargo.toml:1]: I think this crate should be moved to `rust-examples` and should have a readme.

**Review by thealmarty — APPROVED** [review]
Great work Avi. Thank you. I've tested that it works for me. I didn't have valida in my path again and the error message was helpful. Documentation is good. Only thing is that I think the example shouldn't be here, should be in `rust-examples`, as mentioned.

File: src/test_utils.rs [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+                    break;
+                }
+            }
+        }
+    });
+
+    rx
+}
+
+fn set_panic_handler(test: &TestDescAndFn) {
```
[the remark this answers is not in the document]
Comment (Avi-D-coder) [reply src/test_utils.rs:543]: I want to leave it here until a complete hook that halts the vm can is added.

We can add a cutdown hook to entrypoint that just prints at anytime. It won't conflict with this. This code can get removed once the vm process exits correctly.

File: examples/testing/Cargo.toml (hunk 2 above)
Comment (Avi-D-coder) [reply examples/testing/Cargo.toml:1]: I'd like to keep the testing example here with the implementation. That way it stays in sync and as a rust user I would expect to find some examples in the same repo.

Comment (thealmarty) [reply examples/testing/Cargo.toml:1]: I see. That does make sense too. I guess we should also add an example to `rust-examples`?

Comment (Avi-D-coder) [reply examples/testing/Cargo.toml:1]: Yeah, I'll see If can do that tomorrow or next week.

**Comment by Avi-D-coder** [comment]
@morganthomas is this good to merge?

File: src/test_utils.rs (hunk 1 above)
Comment (thealmarty) [reply src/test_utils.rs:421]: @Avi-D-coder are you fixing these? Otherwise good to merge.

**Review by thealmarty — APPROVED** [review]

**Comment by Avi-D-coder** [comment]
I'll make it work with the subcommand before merging, since that requires changes anyway.

**Comment by Avi-D-coder** [comment]
This is ready to merge again.
I force pushed to resolve the conflict so, it should get another review of the force push diff.

File: .cargo/config.toml [hunk 4]
```diff
@@ -1,17 +1,9 @@
...
-linker = "/valida-toolchain/bin/ld.lld"
-rustflags = [
-    "-C", "link-arg=/valida-toolchain/ValidaEntryPoint.o",
-    "-C", "link-arg=--script=/valida-toolchain/valida.ld",
-    "-C", "link-arg=/valida-toolchain/lib/valida-unknown-baremetal-gnu/libc.a",
-    "-C", "link-arg=/valida-toolchain/lib/valida-unknown-baremetal-gnu/libm.a",
-    "-C", "link-arg=--noinhibit-exec",
+[alias]
+valida_config = [
+    "--config", 'build.target="delendum-unknown-baremetal-gnu"',
```
Comment (thealmarty) [inline .cargo/config.toml:3]: delendum has been changed to valida in main

Comment (thealmarty) [reply .cargo/config.toml:3]: everything else in `rust-examples` uses the old `.cargo/config.toml` do we want to make this different?

**Review by morganthomas — APPROVED** [review]

**Pull request merged**